src/odemis/util/angleres.py

```diff
@@ -619,25 +627,66 @@ def ExtractThetaList(data):
     """
-    # For now, return the MD_THETA_LIST for the center wl.
-    # If we eventually interpolate (and crop) the data to the top and bottom
-    # lines, the function could take the wl as an extra argument. Or it'd always
-    # return the values at the center, and another function would map these values
-    # to the top/bottom position for each wavelength.
+    # This should return something like:
+    # NaN, Nan, -78.5°, -30°, -7.5°, 0°, 5°, 15°, 30°, 38°, 43°, NaN
     try:
         wl_list = data.metadata[model.MD_WL_LIST]
-        line_top = data.metadata[model.MD_AR_MIRROR_TOP]
         line_bottom = data.metadata[model.MD_AR_MIRROR_BOTTOM]
+        line_top = data.metadata[model.MD_AR_MIRROR_TOP]
     except KeyError:
         raise ValueError("Metadata required: MD_MIRROR_*, MD_WL_LIST")
-    if len(wl_list) == 0:
-        raise ValueError("Metadata MD_WL_LIST is empty, needs wavelength for each pixel")
+    if len(wl_list) != data.shape[1]:
+        raise ValueError(f"MD_WL_LIST must be the same length as C dimension ({len(wl_list)} != {data.shape[1]})")
+    if len(line_top) != 2 or len(line_bottom) != 2:
+        raise ValueError(f"Expected mirror top & bottom lines to be of length 2, but got {line_top}, {line_bottom}")
+
+    focus_dist = data.metadata.get(model.MD_AR_FOCUS_DISTANCE, AR_FOCUS_DISTANCE)
+    x_max = data.metadata.get(model.MD_AR_XMAX, AR_XMAX)  # optical axis
+    parabola_f = data.metadata.get(model.MD_AR_PARABOLA_F, AR_PARABOLA_F)
+
+    wl = wl_list[data.shape[1] // 2]  # wavelength at the middle (approximately)
+    theta_list, first_px, last_px = _get_theta_list_wl(wl, line_bottom, line_top,
+                                                       focus_dist, x_max, parabola_f,
+                                                       data.shape[0])
+
+    # Replace the angles outside of valid data by NaN
+    theta_list[:first_px] = math.nan
+    theta_list[last_px + 1:] = math.nan
 
+    return theta_list.tolist()
+
+
+def _get_theta_list_wl(wl: float, line_bottom: Tuple[float], line_top: Tuple[float],
+                      focus_dist: float, x_max: float, parabola_f:float,
+                      length: int
+                     ) -> Tuple[numpy.ndarray, int, int]:
+    """
+    Computes the list of theta values given the mirror parameters. More
+    specifically, given that the slit is closed and focused on the center of the
+    mirror and based on the mirror geometry, the detector plane is calculated
+    and the list of angles is derived.
+
+    :param wl > 0: The wavelength, in m, at which these theta values are computed.
+    :param line_bottom: px, px/mm, cf MD_AR_MIRROR_BOTTOM
+    :param line_top: px, px/mm, cf MD_AR_MIRROR_TOP
+    :param focus_dist: m, cf MD_AR_FOCUS_DISTANCE
+    :param x_max: m, cf MD_AR_XMAX
+    :param parabola_f: m, cf MD_AR_PARABOLA_F
+    :param length > 0: the number of angles returned
+    :returns:
```

Comment:
same here

Reply:
Every function in this file is documented with `returns`. I guess it's automatically completed like that by PyCharm. I'm not going to update them. Just the functions I've touched.